### PR TITLE
support EIP 712

### LINKED ISSUE #368: Implement 712 standard
https://github.com/ethereum/EIPs/pull/712

### PR DESCRIPTION
**Pull request by hewigovens**
Fix #368

### CODE REVIEW

File: Trust/Browser/ViewControllers/BrowserViewController.swift [hunk 1]
```diff
@@ -265,6 +265,7 @@ class BrowserViewController: UIViewController {
     }
 }
 
+// MARK: - BrowserNavigationBarDelegate
```
Comment (vikmeup) [inline Trust/Browser/ViewControllers/BrowserViewController.swift]: can we remove a extra comment? not really necessary

File: Trust/Browser/Views/BrowserNavigationBar.swift [hunk 2]
```diff
@@ -32,6 +32,7 @@ class BrowserNavigationBar: UINavigationBar {
         textField.delegate = self
         textField.autocorrectionType = .no
         textField.returnKeyType = .go
+        textField.keyboardType = .URL
```
Comment (vikmeup) [inline Trust/Browser/Views/BrowserNavigationBar.swift]: We can skip keyboardType, there is already one

File: Trust/EtherClient/EtherKeystore.swift [hunk 3]
```diff
@@ -311,6 +311,13 @@ open class EtherKeystore: Keystore {
         return signHash(message.sha3(.keccak256), for: account)
     }
 
+    func signTypedMessage(_ datas: [EthTypedData], for account: Account) -> Result<Data, KeystoreError> {
+        let schemas = datas.map { $0.schemaData }.reduce(Data(), { $0 + $1 }).sha3(.keccak256)
+        let values = datas.map { $0.typedData }.reduce(Data(), { $0 + $1 }).sha3(.keccak256)
+        let combined = (schemas + values).sha3(.keccak256)
+        return signHash(combined, for: account)
```
Comment (vikmeup) [inline Trust/EtherClient/EtherKeystore.swift]: 👍

File: Trust/EtherClient/Types/EthTypedData.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        return Data(bytes: Array(schemaString.utf8))
+    }
+
+    var typedData: Data {
+        if type == "bytes" {
+            return Data(bytes: Array(value.utf8))
+        } else if type == "string" {
+            return Data(bytes: Array(value.utf8))
+        }
+        //FIXME handle other types
```
Comment (vikmeup) [inline Trust/EtherClient/Types/EthTypedData.swift]: how many different types they are?

Comment (nnkken) [reply Trust/EtherClient/Types/EthTypedData.swift]: `bytes1`, `bytes2`, ..., `bytes32`
`uint8`, `uint16`, ..., `uint256`
`int8`, `int16`, ..., `int256`
`address`
`bool`
`string`
`bytes`

See https://github.com/ethereumjs/ethereumjs-abi/blob/master/lib/index.js#L427

Comment (vikmeup) [reply Trust/EtherClient/Types/EthTypedData.swift]: @nnkken thank you for the help! @hewigovens can you take a look?

Comment (hewigovens) [reply Trust/EtherClient/Types/EthTypedData.swift]: Yes, I know these types

**Comment by vikmeup** [comment]
@hewigovens sounds like we have basic functionality in place? 

Are we good to merge or you want to implement rest of the types?

**Pull request merged**